crypto/provider.c

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/err.h>
+#include <openssl/cryptoerr.h>
+#include <openssl/provider.h>
+#include "internal/provider.h"
+
+OSSL_PROVIDER *OSSL_load_provider(OPENSSL_CTX *libctx, const char *name)
+{
+    OSSL_PROVIDER *prov = NULL;
+    OSSL_PROVIDER *created = NULL;
+
+    /* Find it or create it */
+    if ((prov = ossl_provider_find(libctx, name)) == NULL
+        && (prov = created = ossl_provider_new(libctx, name, NULL)) == NULL)
+        return NULL;
+
+    if (!ossl_provider_upref(prov)
+        || !ossl_provider_load(prov)) {
+        ossl_provider_free(created);
+        return NULL;
+    }
+
+    return prov;
+}
+
+int OSSL_unload_provider(OSSL_PROVIDER *prov)
+{
+    ossl_provider_free(prov);
+    return 1;
+}
+
+const OSSL_ITEM *OSSL_get_provider_param_types(OSSL_PROVIDER *prov)
+{
+    return ossl_provider_get_param_types(prov);
+}
+
+int OSSL_get_provider_params(OSSL_PROVIDER *prov, const OSSL_PARAM params[])
+{
+    return ossl_provider_get_params(prov, params);
+}
+
+int OSSL_add_provider(OPENSSL_CTX *libctx,
+                      const char *name, ossl_provider_init_fn *init_fn)
+{
+    if (name == NULL || init_fn == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_ADD_PROVIDER, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    /* Create it */
+    if (ossl_provider_new(libctx, name, init_fn) == NULL)
+        return 0;
+
+    /* It's safely stored in the internal store at this point */
+    return 1;
```

Comment:
But this is only true for built in providers, right?  Those are already partly "loaded" anyway (just not initialized), so I'm not sure I see their constant presence in the store as a bad thing.  And mind you, they *will* get cleaned away the the library context is freed.

Reply:
What if you no longer want the default provider to be in use? E.g. you started using the default provider but some option in the application says "stop using the default provider, I want all my crypto operations to come from some other provider from now on". The design document gives us the flexibility to not load the default provider at all if we want to. We should have also have the ability to unload it if we want to - in the same way that we can for any other provider.

Reply:
To be able to *use* the default provider, it will still have to be fully "loaded" (i.e. initialized), which means there must be a call `OSSL_load_provider("default")`.  That will return the `OSSL_PROVIDER` object for it, which can then be used to called `OSSL_unload_provider`, which makes it "unloaded", i.e. uninitialized.

In other words, when the only reference to a `OSSL_PROVIDER` is the store, that provider should be inactivated...  and I thought I had coded that, but it seems I lost those changes somewhere.  Hang on and I'll reimplement that.

Reply:
The reason behind this inactivate but still present code is to ensure built-in providers are available for loading in much the same way as those in loadable modules.  The problem here is that you can't physically unload built-in code, it will always remain in memory, and also that there's no other way to refer to a built-in provider than to have a name associated with an init function somehow.

The complexity of this code comes down to my wish to have the handling of both kinds of providers look the same to the user.

Reply:
> To be able to _use_ the default provider, it will still have to be fully "loaded" (i.e. initialized), which means there must be a call `OSSL_load_provider("default")`. That will return the `OSSL_PROVIDER` object for it, which can then be used to called `OSSL_unload_provider`, which makes it "unloaded", i.e. uninitialized.

Ah! I had failed to appreciate that you must call both OSSL_add_provider() and OSSL_load_provider() for built-in providers.

Reply:
Ah, good. So back on track then

Reply:
So if the add_provider method is only for setup of built-in providers - maybe the name should reflect this somehow - OSSL_add_builtin_provider, or OSSL_add_loadable_builtin_provider?